src/closure.rs

```diff
@@ -0,0 +1,127 @@
+// TODO: support marshaller.
+
+use std::mem;
+use std::ptr;
+
+use libc::{c_uint, c_void};
+
+use gobject_ffi;
+use translate::{FromGlibPtrFull, FromGlibPtrNone, Stash, StashMut, ToGlibPtr, ToGlibPtrMut, Uninitialized};
+use Value;
+use ToValue;
+
+pub struct Closure {
+    closure: *mut gobject_ffi::GClosure,
+}
+
+impl Closure {
+    pub fn new<F: Fn(&[Value]) -> Option<Value> + 'static>(callback: F) -> Self {
+        unsafe extern "C" fn marshal<F: Fn(&[Value]) -> Option<Value> + 'static>(closure: *mut gobject_ffi::GClosure,
+            return_value: *mut gobject_ffi::GValue, n_param_values: c_uint, mut param_values: *mut gobject_ffi::GValue,
+            invocation_hint: *mut c_void, marshal_data: *mut c_void)
+        {
+            let mut values = vec![];
+            for _ in 0 .. n_param_values {
+                let value: Value = FromGlibPtrNone::from_glib_none(param_values);
+                values.push(value);
+                param_values = param_values.offset(1);
+            }
+            let callback: Box<Box<F>> = Box::from_raw(marshal_data as *mut _);
+            let result = callback(&values);
+            if let Some(result) = result {
+                *return_value = result.into_raw();
+            }
+            for value in values {
+                mem::forget(value);
+            }
+        }
+
+        unsafe {
+            // TODO: get the appropriate size with size_of (not working since the generated
+            // GClosure does not contain every field).
+            let closure = gobject_ffi::g_closure_new_simple(1000, ptr::null_mut());
+            assert_ne!(closure, ptr::null_mut());
+            let callback = Box::new(Box::new(callback));
+            let ptr: *mut Box<F> = Box::into_raw(callback);
+            let ptr: *mut c_void = ptr as *mut _;
+            gobject_ffi::g_closure_set_meta_marshal(closure, ptr, Some(marshal::<F>));
+            Closure {
+                closure: closure,
+            }
+        }
+    }
+
+    pub fn invoke(&mut self, values: &[&ToValue]) -> Option<Value> {
+        // TODO: handle when no return value.
+        let mut value = unsafe { Value::uninitialized() };
+        let values: Vec<_> = values.iter()
+            .map(|v| v.to_value())
+            .collect();
+        let mut gvalues: Vec<gobject_ffi::GValue> = values.iter()
+            .map(|v| unsafe { ptr::read(v.to_glib_none().0) })
+            .collect();
+        unsafe {
+            gobject_ffi::g_closure_invoke(self.to_glib_none_mut().0, value.to_glib_none_mut().0, gvalues.len() as u32,
+                gvalues.as_mut_ptr(), ptr::null_mut());
+        }
+        Some(value)
```

Comment:
This might be unsafe. I think you should a) initialize value with mem::zeroed(), and b) check here if value.g_type == G_TYPE_NONE and in that case return None. What you do here can AFAIU return a Value full of uninitialized memory.